ForestAdmin/forest-express-sequelize
# feat: add typescript declaration file

## Pull Request checklist:

- [x] Create automatic tests
- [x] Review my own code (indentation, syntax, style, simplicity, readability)
- [x] Wonder if you can improve the existing code

## Disclaimer

The is the first attempt of adding Types definition in our repo directly. 

## Motivations

### Observations

The `how to translate your app into typescript` page on the Woodshop is, by far, the second most consulted page. 
Some of our top paying clients using mongoose develops using Typescript

### Problems encountered

We recently have been addressed some issues regarding our Types definition in the community (see this [card](https://app.clickup.com/t/kn5vb3) for more info)

Maintaining Types in DefinitelyTyped has been done using a _community_ approach till now. This leads to some frustration when users upgrade their lianas without updates on their types. Some of our dependencies also have stoped working DefinitelyTyped and integrate their own Types directly in their repos, which makes hard if not impossible to synchronise our own Types. Maintaining Types DefinitelyTyped also leaded to some frustration from us, because updating Types can take up to three weeks to be validated by maintainers (they have a lot to do !). 

## About this version

As stated before, this is the very attempt to add Types in our own repos. Every comments on structure, declarations, scaffolding and naming are welcomed.

## How to test

You can use the following project to test the types: https://github.com/SteveBunlon/fa-blog-ts-postgres
Follow the `README.md` to init the project, then try to use the type (eg. create smart actions, override routes...) and check that auto-completion works as intended.
